### PR TITLE
allow APK auth using assumable identity

### PR DESCRIPTION
**Pull request by imjasonh**
This adds a dep on chainguard.dev/sdk, which is kind of big, unfortunately.

### CODE REVIEW

**Review by cpanato — APPROVED** [review]

**Comment by cpanato** [comment]
i will defer to @mattmoor

File: pkg/apk/auth/chainguard.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	"google.golang.org/api/idtoken"
+)
+
+// NewChainguardIdentityAuth returns an Authenticator that authorizes
+// requests as the given assumeable identity.
+func NewChainguardIdentityAuth(identity string) Authenticator {
+	return authenticator{
+		id:  identity,
+		iss: "https://issuer.enforce.dev", // TODO: make these configurable.
+		aud: "https://apk.cgr.dev",
```
Comment (mattmoor) [inline pkg/apk/auth/chainguard.go:17]: We're not going to be able to test this outside of prod.

File: pkg/apk/auth/chainguard.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+}
+
+type authenticator struct {
+	id, iss, aud string
+}
+
+func (a authenticator) AddAuth(ctx context.Context, req *http.Request) error {
+	ts, err := idtoken.NewTokenSource(ctx, a.iss)
```
Comment (mattmoor) [inline pkg/apk/auth/chainguard.go:26]: This needs to check the Host matches the APK registry before it blindly adds auth

**Review by cpanato — APPROVED** [review]

**Comment by cpanato** [comment]
@imjasonh needs rebase

File: pkg/apk/auth/chainguard.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		aud: audience,
+	}
+}
+
+type authenticator struct {
+	id, iss, aud string
+}
+
+func (a authenticator) AddAuth(ctx context.Context, req *http.Request) error {
+	if req.Host != strings.TrimPrefix(a.aud, "https://") {
```
Comment (mattmoor) [inline pkg/apk/auth/chainguard.go:31]: JFYI I think we expect `aud` to be `apk.cgr.dev` (sans scheme), but this actually allows it to be either, so this is technically fine as-is.

File: pkg/apk/auth/chainguard.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		return err
+	}
+	tok, err := ts.Token()
+	if err != nil {
+		return err
+	}
+	ctok, err := sts.Exchange(ctx, a.iss, a.aud, tok.AccessToken, sts.WithIdentity(a.id))
+	if err != nil {
+		return err
+	}
```
Comment (mattmoor) [inline pkg/apk/auth/chainguard.go:46]: So this doesn't cache at all.  Consider constructing the `TokenSource` in the constructor and wrapping with this to get caching for the token lifetime?

https://pkg.go.dev/golang.org/x/oauth2#ReuseTokenSource

Comment (imjasonh) [reply pkg/apk/auth/chainguard.go]: I've switched to use a `sometimes` like we do for chainctl auth, for consistency.

**Review by mattmoor — APPROVED** [review]

**Comment by cpanato** [comment]
going to merge, thanks!!

**Pull request merged**